### ISSUE TITLE
AWS CloudHSM Test Failures

### ISSUE THREAD

**Issue by howardkitto** [eclipse-keypont/crypto11#19]
```
:~/go/src/github.com/ThalesIgnite/crypto11# go test
        SDK Version: 2.03

C_GenerateKeyPair failed with error CKR_ATTRIBUTE_VALUE_INVALID : 0x00000013
--- FAIL: TestClose (0.16s)
    close_test.go:40: crypto11.GenerateDSAKeyPair: pkcs11: 0x13: CKR_ATTRIBUTE_VALUE_INVALID
2018/12/14 08:49:50 PKCS#11 library already configured
        SDK Version: 2.03

C_GenerateKeyPair failed with error CKR_ATTRIBUTE_VALUE_INVALID : 0x00000013
        SDK Version: 2.03

C_GenerateKeyPair failed with error CKR_ATTRIBUTE_VALUE_INVALID : 0x00000013
        SDK Version: 2.03

C_GenerateKeyPair failed with error CKR_ATTRIBUTE_VALUE_INVALID : 0x00000013
--- FAIL: TestLoginContext (0.64s)
    --- FAIL: TestLoginContext/key_identity_with_login (0.21s)
        crypto11_test.go:57: crypto11.GenerateDSAKeyPair: pkcs11: 0x13: CKR_ATTRIBUTE_VALUE_INVALID
    --- FAIL: TestLoginContext/key_identity_with_expiration (0.21s)
        crypto11_test.go:100: crypto11.GenerateDSAKeyPair: pkcs11: 0x13: CKR_ATTRIBUTE_VALUE_INVALID
    --- FAIL: TestLoginContext/login_context_shared_between_sessions (0.21s)
        crypto11_test.go:134: crypto11.GenerateDSAKeyPair: pkcs11: 0x13: CKR_ATTRIBUTE_VALUE_INVALID
        SDK Version: 2.03

C_GenerateKeyPair failed with error CKR_ATTRIBUTE_VALUE_INVALID : 0x00000013
--- FAIL: TestIdentityExpiration (0.21s)
    crypto11_test.go:178: crypto11.GenerateDSAKeyPair: pkcs11: 0x13: CKR_ATTRIBUTE_VALUE_INVALID
        SDK Version: 2.03

C_GenerateKeyPair failed with error CKR_ATTRIBUTE_VALUE_INVALID : 0x00000013
--- FAIL: TestHardDSA (0.16s)
    dsa_test.go:109: crypto11.GenerateDSAKeyPair: pkcs11: 0x13: CKR_ATTRIBUTE_VALUE_INVALID
2018/12/14 08:49:51 PKCS#11 library already configured

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062

C_SignInit failed with error CKR_KEY_SIZE_RANGE : 0x00000062
--- FAIL: TestHardECDSA (17.49s)
    ecdsa_test.go:108: ECDSA Sign (hash 3): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 4): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 5): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 6): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 7): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 5): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 6): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 3): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 4): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 5): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 6): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 7): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 5): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
    ecdsa_test.go:108: ECDSA Sign (hash 6): pkcs11: 0x62: CKR_KEY_SIZE_RANGE
        SDK Version: 2.03

C_OpenSession failed with error CKR_DEVICE_MEMORY : 0x00000031

C_OpenSession failed with error CKR_DEVICE_MEMORY : 0x00000031

C_OpenSession failed with error CKR_DEVICE_MEMORY : 0x00000031

C_OpenSession failed with error CKR_DEVICE_MEMORY : 0x00000031
--- FAIL: TestHmac (0.12s)
    --- FAIL: TestHmac/HMACSHA1 (0.00s)
        --- FAIL: TestHmac/HMACSHA1/Generate (0.00s)
            hmac_test.go:59: crypto11.GenerateSecretKey: pkcs11: 0x31: CKR_DEVICE_MEMORY
    --- FAIL: TestHmac/HMACSHA1General (0.00s)
        --- FAIL: TestHmac/HMACSHA1General/Generate (0.00s)
            hmac_test.go:59: crypto11.GenerateSecretKey: pkcs11: 0x31: CKR_DEVICE_MEMORY
    --- FAIL: TestHmac/HMACSHA256 (0.00s)
        --- FAIL: TestHmac/HMACSHA256/Generate (0.00s)
            hmac_test.go:59: crypto11.GenerateSecretKey: pkcs11: 0x31: CKR_DEVICE_MEMORY
        SDK Version: 2.03
        SDK Version: 2.03
        SDK Version: 2.03
        SDK Version: 2.03
        SDK Version: 2.03

C_GenerateRandom failed with error CKR_DATA_LEN_RANGE : 0x00000021
--- FAIL: TestRandomReader (0.26s)
    rand_test.go:36: crypto11.PKCS11RandRead.Read: pkcs11: 0x21: CKR_DATA_LEN_RANGE
2018/12/14 08:50:18 PKCS#11 library already configured
        SDK Version: 2.03

C_OpenSession failed with error CKR_DEVICE_MEMORY : 0x00000031

C_OpenSession failed with error CKR_DEVICE_MEMORY : 0x00000031
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x590cdb]

goroutine 114 [running]:
testing.tRunner.func1(0xc0000dde00)
        /usr/local/go/src/testing/testing.go:792 +0x387
panic(0x5cb300, 0x950b60)
        /usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/ThalesIgnite/crypto11.TestHardRSA.func1.2(0xc0000dde00)
        /root/go/src/github.com/ThalesIgnite/crypto11/rsa_test.go:84 +0x2b
testing.tRunner(0xc0000dde00, 0xc000099c20)
        /usr/local/go/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:878 +0x353
exit status 2
FAIL    github.com/ThalesIgnite/crypto11        29.216s
```

**Comment by jvehent** [comment]
Encountered this issue today trying to run `crypto11.GenerateECDSAKeyPairOnSlot(slots[0], []byte(keyName), []byte(keyName), elliptic.P384())` on CloudHSM.

**Comment by optnfast** [comment]
A lot of that looks like certain key types not being supported. The tests can be adjusted to skip them on this provider; https://github.com/ThalesIgnite/crypto11/blob/master/hmac_test.go is an existing example of this.

The `CKR_DEVICE_MEMORY` failures are more puzzling...

**Comment by jvehent** [comment]
The issue I encountered was `C_GenerateKeyPair failed with error CKR_ATTRIBUTE_VALUE_INVALID : 0x00000013` and was caused by duplicate key labels that CloudHSM doesn't appear to permit. Maybe the unit tests also use duplicate labels?

Besides that issue, it appears creating RSA keys is currently broken with CloudHSM. The following code generates rsa keys and works fine with softhsm, but fails with cloudhsm:
```go
package main

import (
        "fmt"
        "log"
        "time"

        "github.com/ThalesIgnite/crypto11"
)

func main() {
        p11Ctx, err := crypto11.ConfigureFromFile("crypto11.config")
        if err != nil {
                log.Fatal(err)
        }
        slots, err := p11Ctx.GetSlotList(true)
        if err != nil {
                log.Fatalf("Failed to list PKCS#11 Slots: %s", err.Error())
        }
        if len(slots) < 1 {
                log.Fatal("No slot found")
        }

        for keySize := 2048; keySize <= 4096; keySize += 256 {
                keyName := []byte(fmt.Sprintf("rsa%d-%d", keySize, time.Now().Unix()))
                fmt.Printf("making %q: ", keyName)
                rsakey, err := crypto11.GenerateRSAKeyPairOnSlot(slots[0], keyName, keyName, keySize)
                if err != nil {
                        fmt.Printf("failed with %v\n", err)
                } else {
                        fmt.Printf("%+v\n", rsakey)
                }
        }
}
```
```
$ ./genrsa
        SDK Version: 2.03
making "rsa2048-1547476079":
C_GenerateKeyPair failed with error CKR_MECHANISM_INVALID : 0x00000070
failed with pkcs11: 0x70: CKR_MECHANISM_INVALID
making "rsa2304-1547476079":
C_GenerateKeyPair failed with error CKR_MECHANISM_INVALID : 0x00000070
failed with pkcs11: 0x70: CKR_MECHANISM_INVALID
making "rsa2560-1547476079":
C_GenerateKeyPair failed with error CKR_MECHANISM_INVALID : 0x00000070
failed with pkcs11: 0x70: CKR_MECHANISM_INVALID
making "rsa2816-1547476079":
C_GenerateKeyPair failed with error CKR_MECHANISM_INVALID : 0x00000070
failed with pkcs11: 0x70: CKR_MECHANISM_INVALID
making "rsa3072-1547476079":
C_GenerateKeyPair failed with error CKR_MECHANISM_INVALID : 0x00000070
failed with pkcs11: 0x70: CKR_MECHANISM_INVALID
making "rsa3328-1547476079":
C_GenerateKeyPair failed with error CKR_MECHANISM_INVALID : 0x00000070
failed with pkcs11: 0x70: CKR_MECHANISM_INVALID
making "rsa3584-1547476079":
C_GenerateKeyPair failed with error CKR_MECHANISM_INVALID : 0x00000070
failed with pkcs11: 0x70: CKR_MECHANISM_INVALID
making "rsa3840-1547476079":
C_GenerateKeyPair failed with error CKR_MECHANISM_INVALID : 0x00000070
failed with pkcs11: 0x70: CKR_MECHANISM_INVALID
making "rsa4096-1547476079":
C_GenerateKeyPair failed with error CKR_MECHANISM_INVALID : 0x00000070
failed with pkcs11: 0x70: CKR_MECHANISM_INVALID
```

**Comment by dmjones** [comment]
@jvehent:  CloudHSM doesn't support the `CKM_RSA_PKCS_KEY_PAIR_GEN` mechanism. See the [list of supported mechanisms](https://docs.aws.amazon.com/cloudhsm/latest/userguide/pkcs11-mechanisms.html). They suggest callers use `CKM_RSA_X9_31_KEY_PAIR_GEN` (which we don't).

I would recommend doing key generation directly via https://github.com/miekg/pkcs11. That allows full control over attributes, mechanisms etc.

Our key generation support is partial and I argue in #15 that we should drop it entirely.

**Comment by dmjones** [comment]
@howardkitto: In the next few weeks I'll be doing some integration testing with CloudHSM, so hopefully can provide more feedback then.

In my experience, CloudHSM is one of the fussier, brittle PKCS#11 implementations. So I'm not surprised that we don't play nicely out of the box.

Apologies for not having an immediate answer, but this issue will be looked into.

**Comment by jvehent** [comment]
Generally speaking, how committed are you to supporting non-thales equipment, like CloudHSM?
My reason for asking is I just noticed the v1 release, which will force a rewrite of our integration. It brings the question of whether we should fork crypto11 to support our cloudhsm use case, or continue tracking upstream.

**Comment by dmjones** [comment]
@jvehent We intend this to be a general purpose library, supporting multiple types of token.

If you have backwards-compatible updates that can address the CloudHSM issue (which alas, I haven't had time to look at yet), we would look to merge them in.

**Comment by jvehent** [comment]
Thanks. I prefer to continue tracking upstream than forking our own version, so if you're interested in maintaining cloudhsm support, I'll move our code to the new v1 interfaces then take a look at this issue again.

**Comment by dmjones** [comment]
That would be great, thanks @jvehent.

**Comment by dmjones** [comment]
I'm finally starting to work on this. Will drop some notes as I go.

First observation is that CloudHSM doesn't support certificate import (source: https://forums.aws.amazon.com/thread.jspa?messageID=809866&#809866). I'll need to document this, plus make it easier to skip those tests in a CloudHSM environment.

**Comment by dmjones** [comment]
This has been addressed in #59. Please reopen this issue (or create a new one) if you encounter problems.
